plonk/src/testing_apis.rs

```diff
@@ -68,7 +68,7 @@ impl<F: Field> From<Challenges<F>> for structs::Challenges<F> {
 /// A wrapper of crate::proof_system::structs::ScalarsAndBases
 #[derive(Debug, Clone)]
 pub struct ScalarsAndBases<E: PairingEngine> {
-    pub base_scalar_map: HashMap<E::G1Affine, E::Fr>,
+    pub base_scalar_map: Vec<(E::G1Affine, E::Fr)>,
```

Comment:
Hmmm. so will this slightly decrease the performance? In the case there are identical bases

Reply:
That's right, this will be worse performance. I was kind of forced to choose this since: 
1. we can't use `HashMap` since that's not available from `alloc`
2. we can't use `BTreeMap` since `G1Affine` doesn't implement `Ord` which is required by [`.entry()`](https://doc.rust-lang.org/alloc/collections/btree_map/struct.BTreeMap.html#method.entry) or any search in the tree.

Can you think of other alternatives?

Reply:
What about https://crates.io/crates/hashbrown/0.12.3